polytracker/grammars.py

```diff
@@ -612,98 +612,95 @@ def __str__(self):
+    with tqdm(unit=" productions", leave=False, desc="extracting a base grammar", total=num_funcs) as t:
+        while func_stack:
+            func = func_stack.pop()
+            t.update(1)
+            prod_name = production_name(func)
+
+            if grammar.start is None and func is entrypoint:
+                grammar.start = Production(
+                    grammar, "<START>", Rule.load(grammar, prod_name)
+                )
 
-        elif isinstance(event, FunctionEntry):
-            prod_name = production_name(event)
-            if event.entrypoint is None:
-                if prod_name not in grammar:
-                    _ = Production(grammar, prod_name)
+            if not func.touched_taint:
+                # do not expand the function if it didn't touch taint
+                _ = Production(grammar, prod_name)
+                t.write(f"skipping {func.function.demangled_name} because it didn't touch taint")
+                t.update(sum(1 for _ in func.calls()))
+                continue
+            elif func.function_entry.entrypoint is None:
+                _ = Production(grammar, prod_name)
             else:
-                rule = Rule(grammar, production_name(event.entrypoint))
+                rule = Rule(grammar, production_name(func.function_entry.entrypoint))
                 if prod_name in grammar:
                     production = grammar[prod_name]
                     if rule not in production:
                         production.add(rule)
                 else:
                     _ = Production(grammar, prod_name, rule)
 
-            if grammar.start is None:
-                grammar.start = Production(
-                    grammar, "<START>", Rule.load(grammar, prod_name)
-                )
+                bbs: List[Optional[BasicBlockEntry]] = list(func.basic_blocks())
+                for bb, next_bb in tqdm(
+                        zip(bbs, bbs[1:] + [None]), leave=False, unit=" basic blocks",
+                        desc=func.function.demangled_name,
+                        delay=1.0,
+                        total=len(bbs)
+                ):
+                    sub_productions: List[Union[Terminal, str]] = [
+                        Terminal(token) for token in bb.consumed_tokens
+                    ]
+
+                    called_function = bb.called_function
+
+                    if called_function is not None:
+                        func_stack.append(called_function)
+                        sub_productions.append(production_name(called_function))
+                        ret = called_function.function_return
+                        if ret is not None:
+                            returning_to = ret.returning_to
+                            if returning_to is not None:
+                                sub_productions.append(f"<{returning_to!s}>")
+                            else:
+                                # TODO: Print warning
```

Comment:
Did you want to add these warnings before merge?

Reply:
A lot of that code gets removed in the forthcoming `early_parser` branch (see #6361), so we'll address it there.